### ISSUE TITLE
faulty select() in activestate perl

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9336]
Migrated from [rt.perl.org#54544](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=54544) (status was 'resolved')

Searchable as RT54544$

**Comment by p5pRT** [comment]
### From perlbug@plan9.de

#### Created by perlbug@plan9.de



select returns an "Unknown error" \(10022\) when it should simply time out&#8203;:

&nbsp;   perl \-e "my $ri = \(chr 0\) x 4; warn select $ri\, undef\, undef\, 0\.1; warn $\!\+0"

in this case\, select should return 0\, to indicate the time out \(this
happens under e\.g\. linux\)\.

instead\, on activestate perl \(windows 2000\)\, I get&#8203;:

&nbsp;   \-10\.1 at \-e line 1\.
&nbsp;   10022 at \-e line 1\.

it seems the empty $ri set triggers this bug\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.10.0:

Configured by Marc Lehmann at Sat Jan 19 07:46:38 CET 2008.

Summary of my perl5 (revision 5 version 10 subversion 0 patch 33007) configuration:
  Platform:
    osname=linux, osvers=2.6.23-1-amd64, archname=amd64-linux
    uname='linux cerebro 2.6.23-1-amd64 #1 smp fri dec 21 12:00:17 utc 2007 x86_64 gnulinux '
    config_args='-Duselargefiles -Dxxxxuse64bitint -Uuse64bitall -Dusemymalloc=n -Dcc=gcc -Dccflags=-DPERL_DONT_CREATE_GVSV_disabled -ggdb -Dcppflags=-DPERL_DONT_CREATE_GVSV_disabled -D_GNU_SOURCE -I/opt/include -Doptimize=-O6 -march=opteron -mtune=opteron -funroll-loops -fno-strict-aliasing -Dcccdlflags=-fPIC -Dldflags=-L/opt/perl/lib -L/opt/lib -Dlibs=-ldl -lm -lcrypt -Darchname=amd64-linux -Dprefix=/opt/perl -Dprivlib=/opt/perl/lib/perl5 -Darchlib=/opt/perl/lib/perl5 -Dvendorprefix=/opt/perl -Dvendorlib=/opt/perl/lib/perl5 -Dvendorarch=/opt/perl/lib/perl5 -Dsiteprefix=/opt/perl -Dsitelib=/opt/perl/lib/perl5 -Dsitearch=/opt/perl/lib/perl5 -Dsitebin=/opt/perl/bin -Dman1dir=/opt/perl/man/man1 -Dman3dir=/opt/perl/man/man3 -Dsiteman1dir=/opt/perl/man/man1 -Dsiteman3dir=/opt/perl/man/man3 -Dman1ext=1 -Dman3ext=3 -Dpager=/usr/bin/less -Uafs -Uusesfio -Uusenm -Uuseshrplib -Dd_dosuid -Dusethreads=undef -Duse5005threads=undef -Duseithreads=undef -Dusemultiplicity=undef -Demail=perl-binary@plan9.de -Dcf_email=perl-binary@plan9.de -Dcf_by=Marc Lehmann -Dlocincpth=/opt/perl/include /opt/include -Dmyhostname=localhost -Dmultiarch=undef -Dbin=/opt/perl/bin -Dxxxusedevel -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DPERL_DONT_CREATE_GVSV_disabled -ggdb -fno-strict-aliasing -pipe -I/opt/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O6 -march=opteron -mtune=opteron -funroll-loops -fno-strict-aliasing',
    cppflags='-DPERL_DONT_CREATE_GVSV_disabled -D_GNU_SOURCE -I/opt/include -DPERL_DONT_CREATE_GVSV_disabled -ggdb -fno-strict-aliasing -pipe -I/opt/include'
    ccversion='', gccversion='4.2.3 20080102 (prerelease) (Debian 4.2.2-5)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-L/opt/perl/lib -L/opt/lib -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-ldl -lm -lcrypt
    perllibs=-ldl -lm -lcrypt
    libc=/lib/libc-2.7.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O6 -march=opteron -mtune=opteron -funroll-loops -fno-strict-aliasing -L/opt/perl/lib -L/opt/lib -L/usr/local/lib'

Locally applied patches:
    MAINT


@INC for perl 5.10.0:
    /root/src/sex
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    .


Environment for perl 5.10.0:
    HOME=/root
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=de_DE.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/root/s2:/root/s:/opt/bin:/opt/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/X11/bin:/usr/games:/usr/local/bin:/usr/local/sbin:/root/pserv:.
    PERL5LIB=/root/src/sex
    PERL5_CPANPLUS_CONFIG=/root/.cpanplus/config
    PERLDB_OPTS=ornaments=0
    PERL_BADLANG (unset)
    PERL_UNICODE=0
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @obra

On Tue May 20 11&#8203;:59&#8203;:58 2008\, perlbug@&#8203;plan9\.de wrote&#8203;:
> 
> This is a bug report for perl from perlbug@&#8203;plan9\.de\,
> generated with the help of perlbug 1\.36 running under perl 5\.10\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> select returns an "Unknown error" \(10022\) when it should simply time
>    out&#8203;:
> 
>    perl \-e "my $ri = \(chr 0\) x 4; warn select $ri\, undef\, undef\, 0\.1;
>    warn $\!\+0"
> 
> in this case\, select should return 0\, to indicate the time out \(this
> happens under e\.g\. linux\)\.
> 
> instead\, on activestate perl \(windows 2000\)\, I get&#8203;:
> 
>    \-10\.1 at \-e line 1\.
>    10022 at \-e line 1\.
> 
> it seems the empty $ri set triggers this bug\.


Bugs in activeperl should be reported to activestate through your 
regular support channel rather than to p5p through perlbug\.  \(I'm aware 
that you filed this bug ~18 months ago and I'm sorry for the delay in 
our reply\.\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@obra - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From schmorp@schmorp.de

On Sat\, Oct 10\, 2009 at 02&#8203;:24&#8203;:35PM \-0700\, Jesse via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> >    \-10\.1 at \-e line 1\.
> >    10022 at \-e line 1\.
> > 
> > it seems the empty $ri set triggers this bug\.
> 
> 
> Bugs in activeperl should be reported to activestate through your 

Activeperl is just an example\, the bug of course is in perl\. Activeperl
is just used as an example \(I \*have\* to use some binary of perl\)\.

\-\- 
&nbsp;                The choice of a       Deliantra\, the free code\+content MORPG
&nbsp;      \-\-\-\-\-==\-     \_GNU\_              http&#8203;://www\.deliantra\.net
&nbsp;      \-\-\-\-==\-\- \_       generation
&nbsp;      \-\-\-==\-\-\-\(\_\)\_\_  \_\_ \_\_\_\_  \_\_      Marc Lehmann
&nbsp;      \-\-==\-\-\-/ / \_ \\/ // /\\ \\/ /      pcg@&#8203;goof\.com
&nbsp;      \-=====/\_/\_//\_/\\\_\,\_/ /\_/\\\_\\

**Comment by p5pRT** [comment]
### From @obra




On Sun\, Oct 11\, 2009 at 04&#8203;:22&#8203;:10AM \+0200\, Marc Lehmann wrote&#8203;:
> On Sat\, Oct 10\, 2009 at 02&#8203;:24&#8203;:35PM \-0700\, Jesse via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > >    \-10\.1 at \-e line 1\.
> > >    10022 at \-e line 1\.
> > > 
> > > it seems the empty $ri set triggers this bug\.
> > 
> > 
> > Bugs in activeperl should be reported to activestate through your 
> 
> Activeperl is just an example\, the bug of course is in perl\. Activeperl
> is just used as an example \(I \*have\* to use some binary of perl\)\.

Activeperl's builds don't always correspond to numbered perl releases\.
Does it fail the same way on a 5\.10\.1 or 5\.11\.0 built from source?

**Comment by p5pRT** [comment]
### From schmorp@schmorp.de

On Sat\, Oct 10\, 2009 at 10&#8203;:57&#8203;:59PM \-0400\, jesse \<jesse@&#8203;fsck\.com> wrote&#8203;:
> Activeperl's builds don't always correspond to numbered perl releases\.
> Does it fail the same way on a 5\.10\.1 or 5\.11\.0 built from source?

No clue\, but the bug is clearly in the source code in win32/win32sck\.c in
5\.10\.0\.

\-\- 
&nbsp;                The choice of a       Deliantra\, the free code\+content MORPG
&nbsp;      \-\-\-\-\-==\-     \_GNU\_              http&#8203;://www\.deliantra\.net
&nbsp;      \-\-\-\-==\-\- \_       generation
&nbsp;      \-\-\-==\-\-\-\(\_\)\_\_  \_\_ \_\_\_\_  \_\_      Marc Lehmann
&nbsp;      \-\-==\-\-\-/ / \_ \\/ // /\\ \\/ /      pcg@&#8203;goof\.com
&nbsp;      \-=====/\_/\_//\_/\\\_\,\_/ /\_/\\\_\\

**Comment by p5pRT** [comment]
### From @obra

On Sat Oct 10 20&#8203;:07&#8203;:42 2009\, schmorp@&#8203;schmorp\.de wrote&#8203;:
> On Sat\, Oct 10\, 2009 at 10&#8203;:57&#8203;:59PM \-0400\, jesse \<jesse@&#8203;fsck\.com> 
wrote&#8203;:
> > Activeperl's builds don't always correspond to numbered perl 
releases\.
> > Does it fail the same way on a 5\.10\.1 or 5\.11\.0 built from source?
> 
> No clue\, but the bug is clearly in the source code in win32/win32sck\.c 
in
> 5\.10\.0\.
> 

It sounds like you have a fairly good handle on what's going wrong\, 
though the original report only mentions the symptom and not the cause\. 
Would you be up for working up a patch and regression test so we can 
make sure it's gone for good?

**Comment by p5pRT** [comment]
### From schmorp@schmorp.de

On Sat\, Oct 10\, 2009 at 08&#8203;:12&#8203;:00PM \-0700\, Jesse via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > No clue\, but the bug is clearly in the source code in win32/win32sck\.c 
> in
> > 5\.10\.0\.
> > 
> 
> It sounds like you have a fairly good handle on what's going wrong\, 
> though the original report only mentions the symptom and not the cause\. 
> Would you be up for working up a patch and regression test so we can 
> make sure it's gone for good?

Ah\, I didn't know the cause back then\, and didn't see my bugreport\, sorry\.

I don't think I have time for a patch\, sorry\. \(time for the patch yes\, but
not for building a windows binary and regression test\)\.

The issue is that win32 select fails with EINVAL unless it has at least
one file descriptor to work with\.

Anyevent&#8203;::Impl&#8203;::Perl works around this like this\, and this should probably
work just fine with perl \(WSAEINVAL is returned only in this case\, or when
the timeout itself is invalid\)&#8203;:

&nbsp;   $fds = CORE&#8203;::select \.\.\.\.\, $wait;

&nbsp;   \[\.\.\.\]

&nbsp;   if \($\! == AnyEvent&#8203;::Util&#8203;::WSAEINVAL\) \{
&nbsp;      CORE&#8203;::select undef\, undef\, undef\, $wait if $wait;
&nbsp;   \}

This uses the faulty workaround code already in place \(which checks for
all three masks missing\, but that's not what is relevant for windows\)\.

I can imagine reacting to EINVAL is good enough\. The "perfect" workaround
would be to count file descriptors and use the existing sleep workaround if
their number is zero\.

\-\- 
&nbsp;                The choice of a       Deliantra\, the free code\+content MORPG
&nbsp;      \-\-\-\-\-==\-     \_GNU\_              http&#8203;://www\.deliantra\.net
&nbsp;      \-\-\-\-==\-\- \_       generation
&nbsp;      \-\-\-==\-\-\-\(\_\)\_\_  \_\_ \_\_\_\_  \_\_      Marc Lehmann
&nbsp;      \-\-==\-\-\-/ / \_ \\/ // /\\ \\/ /      pcg@&#8203;goof\.com
&nbsp;      \-=====/\_/\_//\_/\\\_\,\_/ /\_/\\\_\\

**Comment by p5pRT** [comment]
### From @obra




On Sun\, Oct 11\, 2009 at 05&#8203;:39&#8203;:17AM \+0200\, Marc Lehmann wrote&#8203;:
> On Sat\, Oct 10\, 2009 at 08&#8203;:12&#8203;:00PM \-0700\, Jesse via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > > No clue\, but the bug is clearly in the source code in win32/win32sck\.c 
> > in
> > > 5\.10\.0\.
> > > 
> > 
> > It sounds like you have a fairly good handle on what's going wrong\, 
> > though the original report only mentions the symptom and not the cause\. 
> > Would you be up for working up a patch and regression test so we can 
> > make sure it's gone for good?
> 
> Ah\, I didn't know the cause back then\, and didn't see my bugreport\, sorry\.

No worries\. It's been a while\.

> I don't think I have time for a patch\, sorry\. \(time for the patch yes\, but
> not for building a windows binary and regression test\)\.
 
If you send me a stab at the patch\, I'll happily do the binary build and regression test
dance\. 
 

Best\,
Jesse

**Comment by p5pRT** [comment]
### From schmorp@schmorp.de

On Sat\, Oct 10\, 2009 at 11&#8203;:57&#8203;:35PM \-0400\, jesse \<jesse@&#8203;fsck\.com> wrote&#8203;:
> > I don't think I have time for a patch\, sorry\. \(time for the patch yes\, but
> > not for building a windows binary and regression test\)\.
>  
> If you send me a stab at the patch\, I'll happily do the binary build and regression test
> dance\. 

http&#8203;://data\.plan9\.de/win32\.diff

I opted for the simpler WSAEINVAL route \(minimal changes\)\. The only concern
is select returning WSAEINVAL for invalid time\-outs\, but the existing code
doesn't cope with that either\.

\(I didn't compile the code\, but I hope it is clear what it does\, namely
merey moving the sleep workaround to after select and then checking
whether select returned EINVAL\)\.

\-\- 
&nbsp;                The choice of a       Deliantra\, the free code\+content MORPG
&nbsp;      \-\-\-\-\-==\-     \_GNU\_              http&#8203;://www\.deliantra\.net
&nbsp;      \-\-\-\-==\-\- \_       generation
&nbsp;      \-\-\-==\-\-\-\(\_\)\_\_  \_\_ \_\_\_\_  \_\_      Marc Lehmann
&nbsp;      \-\-==\-\-\-/ / \_ \\/ // /\\ \\/ /      pcg@&#8203;goof\.com
&nbsp;      \-=====/\_/\_//\_/\\\_\,\_/ /\_/\\\_\\

**Comment by p5pRT** [comment]
### From @jandubois

Fixed in http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/e4d771

**Comment by p5pRT** [comment]
@jandubois - Status changed from 'rejected' to 'resolved'

**Comment by p5pRT** [comment]
### From @obra




On Sun\, Oct 11\, 2009 at 04&#8203;:37&#8203;:05PM \+0200\, Marc Lehmann wrote&#8203;:
> On Sat\, Oct 10\, 2009 at 11&#8203;:57&#8203;:35PM \-0400\, jesse \<jesse@&#8203;fsck\.com> wrote&#8203;:
> > > I don't think I have time for a patch\, sorry\. \(time for the patch yes\, but
> > > not for building a windows binary and regression test\)\.
> >  
> > If you send me a stab at the patch\, I'll happily do the binary build and regression test
> > dance\. 
> 
> http&#8203;://data\.plan9\.de/win32\.diff
> 
> I opted for the simpler WSAEINVAL route \(minimal changes\)\. The only concern
> is select returning WSAEINVAL for invalid time\-outs\, but the existing code
> doesn't cope with that either\.
> 
> \(I didn't compile the code\, but I hope it is clear what it does\, namely
> merey moving the sleep workaround to after select and then checking
> whether select returned EINVAL\)\.

Thanks very much\. Based on your patch and problem description\, Jan
added tests and fixed this up in core&#8203;:

http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/e4d771

Best\,
Jesse

**Comment by p5pRT** [comment]
### From @jandubois

On Wed\, 14 Oct 2009\, jesse wrote&#8203;:
> On Sun\, Oct 11\, 2009 at 04&#8203;:37&#8203;:05PM \+0200\, Marc Lehmann wrote&#8203;:
> > On Sat\, Oct 10\, 2009 at 11&#8203;:57&#8203;:35PM \-0400\, jesse \<jesse@&#8203;fsck\.com> wrote&#8203;:
> > > If you send me a stab at the patch\, I'll happily do the binary build and regression test
> > > dance\.
> >
> > http&#8203;://data\.plan9\.de/win32\.diff
> >
> > I opted for the simpler WSAEINVAL route \(minimal changes\)\. The only concern
> > is select returning WSAEINVAL for invalid time\-outs\, but the existing code
> > doesn't cope with that either\.
> >
> > \(I didn't compile the code\, but I hope it is clear what it does\, namely
> > merey moving the sleep workaround to after select and then checking
> > whether select returned EINVAL\)\.
> 
> Thanks very much\. Based on your patch and problem description\, Jan
> added tests and fixed this up in core&#8203;:
> 
> http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/e4d771

Note that the commit doesn't use the EINVAL patch from Marc but implements
what Marc earlier called the "perfect" workaround\, as it wasn't much more
complicated \(we already loop over all file descriptors\)\, and seems a little
more robust&#8203;:

| I can imagine reacting to EINVAL is good enough\. The "perfect" workaround
| would be to count file descriptors and use the existing sleep workaround if
| their number is zero\.

Cheers\,
\-Jan
